### DISCUSSION TITLE
Affiliate product search inlude images

### DISCUSSION THREAD

**Discussion by JojoYou** [etsy/open-api#949] [Q&A]
Hello, I am totally new to Etsy API. I would like to pass my user query to Etsy and I want Etsy to return JSON list of products including their main images. In v2 it worked easily, but in v3 I can't figure out how to do it.

Can someone help, please?

I found some methods to then make an API call for each product, so I can get an image, but I really think this is wasteful.

**Comment by Gareth064** [comment]
Hey @JojoYou. Welcome.

First thing I would ask is, do you know how authentication works in v3? If not then I suggested you get you head around that and get it working before you start looking at calling endpoints. This page will help you understand it. https://developer.etsy.com/documentation/essentials/authentication

If you have that in the bag then here is how you achieve what you want.

You can only get the images of Listings when you API key has been authenticated against a shop. You do that like this:
[getListingsByShop](https://developer.etsy.com/documentation/reference#operation/getListingsByShop) To get all listings associated to a shop. Make sure to "include" images. This will give you an array of the images, and url's to their different sizes.

For shops which aren't authenticate against your API key you can use this, but there is no way to get the images (from what I can see)
[findAllListingsActive](https://developer.etsy.com/documentation/reference#operation/findAllListingsActive) To get all publicly active listings.

**Reply by JojoYou** [reply]
No, I am not aware of how authentication works in v3. :( It looks quite hard to do.

Yea, I use findAllListingsActive, sad that it no longer can return images.
I really don't care from which shop the products came, and as I see you have 1000s of them.  So, to get images I need to do the auth and then choose some of these shops?

**Reply by Gareth064** [reply]
Yeah I'm not sure how you get images of listings when using the public endpoints 🤔.

I might have missed something from the docs, but maybe someone else on here can advise if they have done it before

**Reply by JojoYou** [reply]
Alright, before I used this: https://openapi.etsy.com/v2/listings/active?limit=10&includes=MainImage&api_key='.$_ENV['ETSY_API_KEY'].'&keywords='.$keywords

But, I went through the v3 docs and I also couldn't find a way to get images in public endpoint, but I think for affiliate listings it is necessary to show also images of the products.

In the docs this is written:
We have limited support for the includes query parameter. Our [reference documentation](https://developers.etsy.com/documentation/reference) provides the most up-to-date information, but these are the endpoints that support includes and the resources you can request:

    [getListing](https://developers.etsy.com/documentation/reference#operation/getListing): Images, Shop, User, Translations
    [getListingsByListingIds](https://developers.etsy.com/documentation/reference#operation/getListingsByListingIds): Images, Shop, User, Translations
    [getListingInventory](https://developers.etsy.com/documentation/reference#operation/getListingInventory): Listing
    
    
    So, I wanted to know with what method I can get them.
    
    Still, thank you for the help.

**Reply by Gareth064** [reply]
Those endpoints you mention are all authenticated endpoints.

Only endpoints named "find" are public, as in you just need an API key.

Any which are named "get" require an API key and an Auth Token for the shop you are trying to get listings for.

You can try emailing Etsy Devs to ask how you can get the main image of the listings you can retrieve from the public endpoints. If they give you an answer then please do share what they say 🙂

**Comment by careybaird** [comment]
I am also seeing this... no real workaround that I can see except for making another API request for each individual listing. Hopefully it is added back in

**Reply by Gareth064** [reply]
Email Etsy asking them to add it in. Give them the URL to this GitHub discussion so they can see it's something that people want.

**Reply by JojoYou** [reply]
Hi, I emailed them, they respond:

There are some v3 endpoints to fetch listing images via API which only requires your API key to use here: https://developers.etsy.com/documentation/reference/#tag/ShopListing-Image. But if you're trying to fetch private data, in your shop you need to implement [Oauth Authorization](https://developers.etsy.com/documentation/essentials/authentication).
 
I'd like to share with you our [GitHub issues and discussions page](https://github.com/etsy/open-api/discussions) as you can see some helpful tips for common API concerns and see all the updates or changes to Etsy's API.
 
I hope this information helps. Feel free to contact us back if you have other concerns.




I will reply them now.

**Reply by careybaird** [reply]
Done that. Most of the documentation points to this github repo if you need help, but I did find the email developers@etsy.com

**Reply by careybaird** [reply]
Thanks @JojoYou 

That didn't really answer our problem.. I think they are looking at it from a shop owner point of view, instead of as an affiliate

**Reply by JojoYou** [reply]
Seems like it.

These were my emails:

1st email:
Hello Etsy developers, I am very new user of your API, I would like to

use it for affiliate using public endpoins. In version 2 I used this
query:
https://openapi.etsy.com/v2/listings/active?limit=10&includes=MainImage&api_key='.$_ENV['ETSY_API_KEY'].'&keywords='.$keywords

It includes MAIN IMAGES


2nd email:
I somehow accidentally clicked send, I don't know how that happened. I wanted to ask if there is a way on how to get product images from find - public endpoints.

**Comment by careybaird** [comment]
Still waiting on this one, Etsy api support weren't much help

**Reply by AhamSammich** [reply]
I've been able to get all listings with images using these unauthenticated endpoints.

***[findAllActiveListingsByShop](https://developers.etsy.com/documentation/reference#operation/findAllActiveListingsByShop)***
The data from this endpoint does not include images, but I pulled the listing_ids from the response into an array.

***[getListingsByListingIds](https://developers.etsy.com/documentation/reference#operation/getListingsByListingIds)***
This endpoint can take the array of listing_ids and 'includes=Images' as query parameters.

This gets you all listings with images with only 2 requests and no oauth token required.

**Reply by Gareth064** [reply]
I didn't realize the `getListingsByListingIds` was a public endpoint. Ages ago, someone from Etsy said that endpoints beginning with `get` would require an access token, and any beginning with `find` would just need an api key.

Going to have to recheck all these endpoints again to double check them

**Reply by piwikbugs** [reply]
I'm having difficulties getting this to work... I can't find the proper way to format this and I'm getting a "Expected int value for 'listing_id' (got string)" error.

How can I include an array of listings as an integer?

`$etsyUrl = "https://openapi.etsy.com/v3/application/listings/batch&listing_ids=12341234,34563456,465785678";`

doesn't work unfortunately.

Thanks so much for your time!

**Reply by emergingdzns** [reply]
@piwikbugs Your URL is malformed. `batch&listing_ids=` should be `batch?listing_ids=`. Then you can append `&includes=Images` to have it return the images.

**Comment by piwikbugs** [comment]
Oh my... what a ridiculous typo... Even after all these years :-)

Thanks so much! Works now, and I can finish my script!
